Add doc for mutexlock / mutexunlock / protect (#2132)
neuronsimulator/nrn: docs/python/modelspec/programmatic/mechanisms/nmodl2.rst

~~~~~~~

Description:
    .. code-block::

        NEURON {
            GLOBAL var
        }

        BREAKPOINT {
            PROTECT var = var + 1
        }

    Mod files that assign values to :ref:`GLOBAL` variables are not considered
    thread safe. If the mod file is using the ``GLOBAL`` as a counter, prefix
    the offending assignment statements with the ``PROTECT`` keyword so that
    multiple threads do not attempt to update the value at the same time

>     (race condition). If the mod file is using the ``GLOBAL`` essentially as

Reviewer: ```suggestion
    (leading to a race condition). If the mod file is using the ``GLOBAL`` essentially as
```

    a file scope :ref:`LOCAL` along with the possibility of passing values back
    to hoc in response to calling a :ref:`PROCEDURE`, use the :ref:`THREADSAFE` keyword
    in the :ref:`NEURON` block to automatically treat those GLOBAL variables
    as thread specific variables. NEURON assigns and evaluates only
    the thread 0 version and if :ref:`FUNCTION` and ::ref:`PROCEDURE` are called from
    Python, the thread 0 version of these globals are used.

    .. note::

>         Protect share the same mutex than :ref:`MUTEXLOCK / MUTEXUNLOCK`

Reviewer: ```suggestion
        `PROTECT` shares the same mutex as :ref:`MUTEXLOCK / MUTEXUNLOCK`
```

    ``TODO``: Add existing example mod file (share/demo/release/mcna.mod)

MUTEXLOCK / MUTEXUNLOCK
~~~~~~~~~~~~~~~~~~~~~~~

Description:
    .. code-block::

        LOCAL factors_done

        INITIAL {
            MUTEXLOCK
            if (factors_done == 0) {
                  factors_done = 1
                  factors()
            }
            MUTEXUNLOCK
        }

        PROCEDURE factors() {
            : ...
        }

>     ``MUTEXLOCK`` and ``MUTEXUNLOCK`` are the two directives to handle threadsafety inside zones.

Reviewer: ```suggestion
    ``MUTEXLOCK`` and ``MUTEXUNLOCK`` are two related directives to handle thread-safety inside zones.
```

>     Directive respectively lock and unlock the unique global mutex (one per mechanismi's instance).

Reviewer: ```suggestion
    The two directives respectively lock and unlock the unique global mutex (one per mechanism's instance).
```

~~~~~~~

Description:
    .. code-block::

        NEURON {
            GLOBAL var
        }

        BREAKPOINT {
            PROTECT var = var + 1
        }

    Mod files that assign values to :ref:`GLOBAL` variables are not considered
    thread safe. If the mod file is using the ``GLOBAL`` as a counter, prefix
    the offending assignment statements with the ``PROTECT`` keyword so that
    multiple threads do not attempt to update the value at the same time
    (leading to a race condition). If the mod file is using the ``GLOBAL`` essentially as
    a file scope :ref:`LOCAL` along with the possibility of passing values back
    to hoc in response to calling a :ref:`PROCEDURE`, use the :ref:`THREADSAFE` keyword
    in the :ref:`NEURON` block to automatically treat those ``GLOBAL`` variables
    as thread specific variables. NEURON assigns and evaluates only
    the thread 0 version and if :ref:`FUNCTION` and ::ref:`PROCEDURE` are called from
    Python, the thread 0 version of these globals are used.

    .. note::
        ``PROTECT`` shares the same mutex as :ref:`MUTEXLOCK / MUTEXUNLOCK`

>     ``TODO``: Add existing example mod file (share/demo/release/mcna.mod)

Reviewer 2: This `TODO` could also be removed since you added the snippet of code above I think

MUTEXLOCK / MUTEXUNLOCK
~~~~~~~~~~~~~~~~~~~~~~~

Description:
    .. code-block::

        LOCAL factors_done

        INITIAL {
            MUTEXLOCK
            if (factors_done == 0) {
                  factors_done = 1
                  factors()
            }
            MUTEXUNLOCK
        }

        PROCEDURE factors() {
            : ...
        }

    ``MUTEXLOCK`` and ``MUTEXUNLOCK`` are two related directives to handle thread-safety inside zones.
    The two directives respectively lock and unlock the unique global mutex (one per mechanism's instance).

    .. note::
        They share the same mutex than :ref:`PROTECT`.

>     ``TODO``: Add description and existing example mod file (share/examples/nrniv/nmodl/cadif.mod)

Reviewer 2: Same as above
